man/man1/pcpustat.1
.\" man page for the pcpustat utility.
.Dd September 26, 2010
.Dt PCPUSTAT 1
.Os
.Sh NAME
.Nm pcpustat
.Nd per-CPU usage statistics
.Sh SYNOPSIS
.Nm
.Op Fl ahinqstux
.Op Fl c Ar count
.Op Fl d Ar delimiter
.Op Fl p Ar cpu
.Op Fl w Ar wait
.Sh DESCRIPTION
The pcpustat utility provides per-CPU usage statistics.  This information is computed using the same algorithm as in top(1) with the -P option, but the output format is more similar to that of iostat(8) so that it can be piped and reformatted.
.Sh OPTIONS
.Bl -tag -width indent
.It Fl a, all
Include all usage statistics (same as -instu).  This is the default if no specific statistics are selected.
.It Fl c, count Ar count
Sample statistics the specified number of times.  If not specified, it defaults to 1 if -w is not specified, otherwise forever.
.It Fl d, delim Ar delimiter
Delimit columns with delimiter instead of justifying them with spaces.
.It Fl h, help
Print the list of options and exit.
.It Fl i, idle
Include idle time in the statistics.
.It Fl n, nice
Include nice time in the statistics.
.It Fl p, cpu Ar cpu
Select processor cpu [0-n].  You may repeat the switch to select more than one CPU.  The default is to include all CPUs.
.It Fl q, quiet
Print no header, only the data.
.It Fl s, system
Include system time in the statistics.
.It Fl t, interrupt
Include interrupt time in the statistics.
.It Fl u, user
Include user time in the statistics.
.It Fl w, wait Ar wait
Pause the specified number of seconds between updates.  The default is 1.
.It Fl x, not
Report the percentage of time not in each state.  If specified, each percentage is subtracted from 100.  The header (unless the -q option is selected) displays a "!" before each heading.
.Sh OUTPUT FORMAT
If the 
.Op Fl q
option is not specified, the first two lines of output will contain a header.
.Pp
The first line lists the CPUs, by number.  If the options
specify only one column of information per CPU, then this line will only contain the CPU numbers (starting with 0), otherwise it will contain
the string "cpu " followed by the number.  This heading will be right-justified over the columns that apply to that CPU, unless the
.Op Fl d
option is specified (in which case the headers are also delimited).
.Pp
The second header line labels the columns within the output for each CPU.  These labels can be any of the following:
.Pp
.Bl -tag -width indent -compact
.It \&us
% of cpu time in user mode
.It \&ni
% of cpu time in user mode running niced processes
.It \&sy
% of cpu time in system mode
.It \&in
% of cpu time in interrupt mode
.It \&id
% of cpu time in idle mode
.El
.Pp
If the
.Op Fl x
option was specified, then each of the above is prefixed by '!'.
.Pp
The remaining lines of output will be the percentage of time each CPU spent in each of the designated states since the previous line (or since pcpustat started, in the case of the first line).
.Pp
If the
.Op Fl d
option is specified, then the string specified as its argument will occur between each column.  Otherwise, enough spaces will be used to
right-align each column under its heading.
.Sh SEE ALSO
top(1), iostat(8), sysctl(3), sysctl(8)
.Sh AUTHOR
Sterling (Chip) Camden <sterling@camdensoftware.com>
.Sh SITE
http://chipstips.com/?tag=c_pcpustat
.Sh BUGS
The pcpustat utility can only report on CPUs 0-n, where n is one less than the number of bits in a long.


pcpustat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/param.h>
#include <sys/errno.h>
#include <sys/resource.h>
#include <sys/sysctl.h>

#ifndef CPUSTATES
#define CPUSTATES	5	/* OSX doesn't define this */
#endif

static const char* what_string="@(#)pcpustat 1.4";

/* Bit flags for what stats to include: */

#define STAT_USER	0x01
#define STAT_NICE	0x02
#define STAT_SYSTEM	0x04
#define STAT_INT	0x08
#define STAT_IDLE	0x10
#define STAT_ALL	0x1F


/* Copied from /usr/src/usr.bin/top/machine.c */
#define GETSYSCTL(name, var) getsysctl(name, &(var), sizeof(var))
static void getsysctl(const char *name, void *ptr, size_t len);
static size_t getsysctllen(const char *name);

struct opthelp {
    char *argname;
    char *description;
};

int main(int ac, char **av)
{
    int c, option_index, stats=0, count=-1, wait=0, cpu, ncpu, quiet=0, not=0;
    char *delim=NULL;
    size_t state_size;
    long cpus=0;
    long *cpu_prev, *cpu_curr;
    struct option options[] = {
      {"all", 0, 0, 'a'},
      {"count", 1, 0, 'c'},
      {"delim", 1, 0, 'd'},
      {"help", 0, 0, 'h'},
      {"idle", 0, 0, 'i'},
      {"nice", 0, 0, 'n'},
      {"cpu", 1, 0, 'p'},
      {"quiet", 0, 0, 'q'},
      {"system", 0, 0, 's'},
      {"interrupt", 0, 0, 't'},
      {"user", 0, 0, 'u'},
      {"wait", 1, 0, 'w'},
      {"not", 0, 0, 'x'}
    };
    struct opthelp help[] = {
      {0,"include all usage statistics (-instu)"},
      {"count","repeat count times (default = forever if wait is specified)"},
      {"delimiter","separate columns with delimiter instead of justifying with spaces"},
      {0,"print this list and exit"},
      {0,"include idle time"},
      {0,"include nice time"},
      {"cpu","select processor cpu [0-n] (default = all CPUs)"},
      {0,"print no header"},
      {0,"include system time"},
      {0,"include interrupt time"},
      {0,"include user time"},
      {"wait","pause wait seconds between updates (default = 1)"},
      {0,"report percentage not in each state"}
    };

    GETSYSCTL("hw.ncpu", ncpu);

    if (ncpu > sizeof(long)*8)
      ncpu = sizeof(long)*8;	/* We're using a bit per CPU */

    while ((c = getopt_long(ac, av, "ac:d:hinp:qstuw:x", options, &option_index)) >= 0)
    {
        switch (c)
        {
	    case 'a':
	        stats = STAT_ALL;
		break;

	    case 'c':
		count = strtol(optarg, NULL, 10);
	        break;

	    case 'd':
	    	delim = malloc(strlen(optarg)+1);
		strcpy(delim, optarg);
		break;

            case 'h':
	    {
	        int n, nopts = (sizeof(options) / sizeof(struct option));
		printf("%s\nusage: pcpustat [-", what_string+4);
		for (n = 0; n < nopts; n++)
		  if (!(help[n].argname))
		    printf("%c", options[n].val);
		printf("]");
		for (n = 0; n < nopts; n++)
		  if (help[n].argname)
		    printf(" [-%c %s]", options[n].val, help[n].argname);
		printf("\n");
		for (n = 0; n < nopts; n++)
		    printf("  -%c, --%-10s\t%s\n", options[n].val, options[n].name,
		    	help[n].description);
	    	exit(1);
	    }
	    break;

	    case 'i':
	        stats |= STAT_IDLE;
		break;

	    case 'n':
		stats |= STAT_NICE;
	    	break;

	    case 'p':
	    {
	        cpu = strtol(optarg, NULL, 10);
		if ((cpu < 0) || (cpu > (ncpu-1)))
		{
		    fprintf(stderr, "error: -p specifies invalid CPU number, range 0-%d\n", ncpu-1);
		    exit(1);
		}
		cpus |= 1 << cpu;
	    }
	    break;

	    case 'q':
	        quiet = 1;
		break;

	    case 's':
	    	stats |= STAT_SYSTEM;
		break;

	    case 't':
	        stats |= STAT_INT;
		break;

	    case 'u':
		stats |= STAT_USER;
	    	break;

	    case 'w':
		wait = strtol(optarg, NULL, 10);
	    	break;

	    case 'x':
	        not = 1;
		break;
        }
    }

    if (!stats)
        stats = STAT_ALL;

    if (!cpus)
        cpus = -1;	/* All CPUs */

    if (wait < 1)
    {
	if (count < 1)
            count = 1;
	wait = 1;
    }

    if (!quiet)
    {
	char *head, fmt[5];
	int i;

	if (delim)
	{
	    head = malloc((strlen(delim)+3)*5+1);
	    char *p = head;
	    int cnt = 0;
	    #define ADD_HEAD(s) { \
				    if (not) *p++ = '!';\
				    sprintf(p,"%s%s",s,delim);\
				    p+=strlen(p);\
				    cnt += 1;\
				  }
	    if (stats & STAT_USER)
	      ADD_HEAD("us");
	    if (stats & STAT_NICE)
	      ADD_HEAD("ni");
	    if (stats & STAT_SYSTEM)
	      ADD_HEAD("sy");
	    if (stats & STAT_INT)
	      ADD_HEAD("in");
	    if (stats & STAT_IDLE)
	      ADD_HEAD("id");
	    #undef ADD_HEAD
	    *(p-=strlen(delim)) = 0;		/* Overwrite last delimiter */
	    cnt--;
	    for (cpu = 0; cpu < ncpu; cpu++)
	    {
	      int dc;
	      if (cpu > 0)
	        printf(delim);
	      printf("cpu %d", cpu);
	      for (dc=0; dc < cnt; dc++)
	        printf(delim);
	    }
	    printf("\n");
	}
	else
	{
	    head = malloc(21);
	    sprintf(head, "%s%s%s%s%s",
		(stats & STAT_USER) ? "  us" : "",
		(stats & STAT_NICE) ? "  ni" : "",
		(stats & STAT_SYSTEM) ? "  sy" : "",
		(stats & STAT_INT) ? "  in" : "",
		(stats & STAT_IDLE) ? "  id" : "");
	    if (not)
	      for (i=1; i<strlen(head); i+=4)
		head[i] = '!';
	    sprintf(fmt, "%%%ds", (int)strlen(head));

	    for (cpu = 0; cpu < ncpu; cpu++)
	      if (cpus & (1 << cpu))
	      {
		char str[8];
		if (strlen(head) > 5)
		  sprintf(str, "cpu %d", cpu);
		else
		  sprintf(str, "%d", cpu);
		printf(fmt, str);
	      }
	    printf("\n");
	}
	for (cpu = 0; cpu < ncpu; cpu++)
	  if (cpus & (1 << cpu))
	  {
	    if (delim && (cpu > 0))
	      printf(delim);
	    printf(head);
	  }
	printf("\n");
	free(head);
    }

    state_size = getsysctllen("kern.cp_times");
    cpu_prev = malloc(state_size);
    cpu_curr = malloc(state_size);
    getsysctl("kern.cp_times", cpu_prev, state_size);

    for (;count; count--)	/* Negative = forever */
    {
	long *prev, *curr;
	int first = 1;

	if (wait > 0)
            sleep(wait);
	getsysctl("kern.cp_times", cpu_curr, state_size);

	for (cpu=0, prev=cpu_prev, curr=cpu_curr;
	     cpu < ncpu;
	     cpu++, prev += CPUSTATES, curr += CPUSTATES)
	  if (cpus & (1 << cpu))
	  {
	    long diff[CPUSTATES], n, h, total=0;

	    for (n=0; n < CPUSTATES; n++)
	    {
	        diff[n] = curr[n] - prev[n];
		total += diff[n];
	    }
	    h = total / 2;
	    for (n=0; n < CPUSTATES; n++)
	      if (stats & (1 << n))
	      {
	        int pct = (int)((diff[n] * 100 + h) / total);
		if (not)
		  pct = 100 - pct;
		if (delim)
		{
		    if (first)
		       first = 0;
		    else
		       printf(delim);
		    printf("%d", pct);
		}
		else
		{
	            printf(" %3d", pct);
		}
	      }
	  }
	printf("\n");
	fflush(stdout);

	memmove(cpu_prev, cpu_curr, state_size);
    }

    free(cpu_prev);
    free(cpu_curr);
    if (delim)
      free(delim);
    return(0);
}

/* Copied from /usr/src/usr.bin/top/machine.c */

static void
getsysctl(const char *name, void *ptr, size_t len)
{
	size_t nlen = len;

	if (sysctlbyname(name, ptr, &nlen, NULL, 0) == -1) {
		fprintf(stderr, "pcpustat: sysctl(%s...) failed: %s\n", name,
		    sys_errlist[errno]);
		exit(23);
	}
}

static size_t
getsysctllen(const char *name)
{
	size_t len = 0;
	if (sysctlbyname(name, NULL, &len, NULL, 0) == -1) {
		fprintf(stderr, "pcpustat: sysctl(%s...) failed: %s\n", name,
		    sys_errlist[errno]);
		exit(23);
	}
	return len;
}
